panther_manager/include/panther_manager/plugins/shutdown_hosts_node.hpp

```diff
@@ -0,0 +1,150 @@
+#ifndef PANTHER_MANAGER_SHUTDOWN_HOSTS_NODE_HPP_
+#define PANTHER_MANAGER_SHUTDOWN_HOSTS_NODE_HPP_
+
+#include <algorithm>
+#include <memory>
+#include <set>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <behaviortree_cpp/action_node.h>
+#include <behaviortree_cpp/basic_types.h>
+#include <behaviortree_cpp/tree_node.h>
+
+#include <ros/ros.h>
+
+#include <panther_manager/plugins/shutdown_host.hpp>
+
+namespace panther_manager
+{
+
+class ShutdownHosts : public BT::StatefulActionNode
+{
+public:
+  explicit ShutdownHosts(const std::string & name, const BT::NodeConfig & conf)
+  : BT::StatefulActionNode(name, conf)
+  {
+    node_name_ = ros::this_node::getName();
+  }
+
+  virtual ~ShutdownHosts() = default;
+
+  // method to be implemented by user
+  virtual void update_hosts(std::vector<std::shared_ptr<ShutdownHost>> & hosts) = 0;
+
+  // method that can be overriden by user
+  virtual BT::NodeStatus post_process()
+  {
+    // return success only when all hosts succeded
+    if (failed_hosts_.size() == 0) return BT::NodeStatus::SUCCESS;
+    return BT::NodeStatus::FAILURE;
+  }
+
+  std::string get_node_name() const { return node_name_; }
+  std::vector<std::size_t> const get_failed_hosts() { return failed_hosts_; }
+
+private:
+  int check_host_index_ = 0;
+  std::string node_name_;
+  std::vector<std::shared_ptr<ShutdownHost>> hosts_;
+  std::vector<std::size_t> hosts_to_check_;
+  std::vector<std::size_t> skipped_hosts_;
+  std::vector<std::size_t> succeeded_hosts_;
+  std::vector<std::size_t> failed_hosts_;
+
+  BT::NodeStatus onStart()
+  {
+    update_hosts(hosts_);
+    remove_duplicate_hosts(hosts_);
+    if (hosts_.size() <= 0) {
+      ROS_ERROR("[%s] Hosts list is empty! Check configuration!", node_name_.c_str());
+      return BT::NodeStatus::FAILURE;
+    }
+    for (std::size_t i = 0; i < hosts_.size(); i++) {
+      hosts_to_check_.push_back(i);
+    }
+    return BT::NodeStatus::RUNNING;
+  }
+
+  BT::NodeStatus onRunning()
+  {
+    if (hosts_to_check_.size() <= 0) return post_process();
+
+    if (check_host_index_ >= hosts_to_check_.size()) check_host_index_ = 0;
```

Comment:
```suggestion
    if (hosts_to_check_.size() <= 0) {
      return post_process();
    }

    if (check_host_index_ >= hosts_to_check_.size()) {
      check_host_index_ = 0;
    }
```
I would avoid one line if statements for the sake of readability